### DISCUSSION TITLE
Is there a way to crop the unused space of a rectangle in pymupdf ?

### DISCUSSION THREAD

**Discussion by smithah** [pymupdf/PyMuPDF#3634] [Looking for help]
The text being written into a rectangle leave some space at the end, need to know if there is a way to crop the unused spaces in a rectangle, so there is no unnecessay white spaces between adjacent lines

**Comment by JorjMcKie** [comment]
Depending on your insertion method, there are different way how the unused rectangular stripe at the rectangle's bottom is reported back. See documentation for [insert_textbox](https://pymupdf.readthedocs.io/en/latest/page.html#Page.insert_textbox), [fill_textbox](https://pymupdf.readthedocs.io/en/latest/textwriter.html#TextWriter.fill_textbox) and [insert_htmlbox](https://pymupdf.readthedocs.io/en/latest/page.html#Page.insert_htmlbox).
Where in reality that empty stripe is located must be interpreted by you depending on the rotation angle you chose for the text - IAW if you chose 180° then the rectangle will have been filled from bottom to top and the empty stripe is located at the top accordingly.
Depending on this information, compute the placement of your next text rectangle.

All three methods also have ways to control the inter-line distance.

**Comment by smithah** [comment]
Thanks for the response will check on them

**Reply by smithah** [reply]
even after calculating the text width i see space in the insert_htmlbox

![image](https://github.com/pymupdf/PyMuPDF/assets/28997203/e3da6ad7-0121-4794-945c-9913aa398e0e)

Its leaving quite and amount of space in the rectangle

![image](https://github.com/pymupdf/PyMuPDF/assets/28997203/70b42c75-c687-4499-9c72-621994a0bea7)


Following were done :

1) Calculate the width of the text with the PIL and `sum(font1.char_lengths(**bidi_text**+" ",fontsize=font_size))`
but still there are gaps in the rectangle formed as seen in the image above, the next line starts after that rectangle.

2) Reduced the fontsize to an optimal value 

3) Tried rotation but it rotates the text altogether within the textbox.

**Comment by smithah** [comment]
https://github.com/pymupdf/PyMuPDF/discussions/3267

Trying this approach, if it solves the issue faced

**Reply by smithah** [reply]
But it's extremely slow using insert_htmlbox

**Reply by JorjMcKie** [reply]
Of course it is slow - it **_must_** be!
As always: comfort has a price - see this [blog article](https://artifex.com/blog/mastering-pdf-text-with-pymupdfs-insert-htmlbox-what-you-need-to-know).

If you don't like the speed or don't need the comfort - simply do not use it!

**Reply by smithah** [reply]
Is there a way to find empty space exists in the rectangle after text is inserted? So the text font size can be increased till it exactly fits bbox width.

**Reply by smithah** [reply]
Referring this 
https://github.com/pymupdf/PyMuPDF/issues/1146
